### ISSUE TITLE
Compressing carriage-return in log output

### ISSUE THREAD

**Issue by farid-fari** [Nukesor/pueue#184]
**Is your feature request related to a problem? Please describe.**
When I run an `ffmpeg` command, and then check the log, the initial output is displayed correctly but then the actual progress is a succession of lines and then backscape escape characters, which makes them flicker quickly at the bottom of the log output then disappear.

**Describe the solution you'd like**
Not really sure how to make something clean and standard. One solution is detect the command is ffmpeg and thus handle the progress output accordingly (remove all backscape characters and keep only the last progress line), but something more general could be to remove these escape characters completely at the cost of potentially broken output.

**Comment by farid-fari** [comment]
First of all thanks for the detailed answer. It's true doing it only for ffmpeg would be quite a dirty fix and hard to maintain. I only suggested that because I thought it would be the only complete solution.
It's true the caret handling solution would probably be the most complete, because the log would look like what the user would see if they were running the command themselves in a shell outside of pueue.
I suppose another alternative would be an option that would handle these types of cases: maybe one that hides everything from the previous \n everytime it sees a \r, but maybe there might be other relevant alternatives?

**Comment by Nukesor** [comment]
I'm thinking about writing a small mini-crate that does this exact thing, while exposing a generic `Read` interface :D 
It sounds like a good opportunity to dive into low-level buffer reading and stream handling :)

This might take a while, though, since I want to do and test this properly.

**Comment by farid-fari** [comment]
Sure no problem, it's certainly not urgent or absolutely necessary.
A workaround meanwhile is piping `pueue log` into `less`, since it escapes the `\r` characters.

**Comment by Nukesor** [comment]
I looked into this, and it turns out that doing this in a high-performance way is pretty tricky.
I'm not really satisfied with any solution that I came up with.

For anyone interested in this issue:

I experimented with a BufReader like struct, since we have to stream this.
The implementation should be able to stream a Gigabyte of data in one or two seconds.

Problems:
1. Wrong handling, if lines get longer than the curren buffer size
2. We would have to shift the buffer all the time, since we want to have as much lookup as possible (We find a newline, return that content, shift the remaining buffer to the left and refill)
3. It's possible to detect `\n` and `\r` characters in a bytestream, due to their unique byte constellation.
     However, it's not simple to replace existing UTF-8 bytes due to multi-byte UTF-8 characters.
     The implementation would have to be character-aware, which makes this whole thing a lot more complex and imperformant...
     
This might still be possible, but I'm currently not in the mood to do this myself :D
Might be a cool project, if you have the time, though ;)
